sdk/sdktypes/session_log_record_describe.go

```diff
@@ -0,0 +1,285 @@
+package sdktypes
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"go.autokitteh.dev/autokitteh/internal/kittehs"
+	modulev1 "go.autokitteh.dev/autokitteh/proto/gen/go/autokitteh/module/v1"
+)
+
+type SessionLogRecordDescribeOptions struct {
+	IncludeProcessID bool
+	IncludeTime      bool
+	TimeFormat       string
+	Indent           string
+	UnwrapValues     bool
+}
+
+func (o SessionLogRecordDescribeOptions) indent(w io.Writer) io.Writer {
+	indent := o.Indent
+	if indent == "" {
+		indent = "  "
+	}
+
+	return kittehs.NewIndentedStringWriter(w, indent)
+}
+
+var DefaultSessionLogRecordDescribeOptions = &SessionLogRecordDescribeOptions{
+	IncludeTime:  true,
+	TimeFormat:   time.RFC3339,
+	Indent:       "  ",
+	UnwrapValues: true,
+}
+
+func (r SessionLogRecord) ToString() string { return r.Describe(nil) }
+
+func (r SessionLogRecord) Describe(opts *SessionLogRecordDescribeOptions) string {
+	if opts == nil {
+		opts = DefaultSessionLogRecordDescribeOptions
+	}
+
+	m := r.read()
+
+	b := &strings.Builder{}
+
+	if opts.IncludeProcessID && m.ProcessId != "" {
+		fmt.Fprintf(b, "%s ", m.ProcessId)
+	}
+
+	if opts.IncludeTime && m.T != nil {
+		fmt.Fprintf(b, "[%s] ", r.Timestamp().Format(opts.TimeFormat))
+	}
+
+	switch {
+	case m.Print != nil:
+		fmt.Fprintf(b, "Print: %s\n", m.Print.Text)
+	case m.StopRequest != nil:
+		fmt.Fprintf(b, "Stop Request: %s\n", m.StopRequest.Reason)
+	case m.State != nil:
+		fmt.Fprint(b, "State: ")
+		describeState(opts, b, r.GetState())
+	case m.CallSpec != nil:
+		fmt.Fprint(b, "Call Spec: ")
+		describeCallSpec(opts, b, r.GetCallSpec())
+	case m.CallAttemptStart != nil:
+		fmt.Fprintf(b, "Call Attempt Start: Attempt #%d\n", m.CallAttemptStart.Num)
+	case m.CallAttemptComplete != nil:
+		fmt.Fprintln(b, "Call Attempt Complete")
+		w := opts.indent(b)
+		if m.CallAttemptComplete.IsLast {
+			fmt.Fprintln(w, "Final")
+		} else {
+			fmt.Fprintf(w, "Retry in %s\n", m.CallAttemptComplete.RetryInterval)
+		}
+		if m.CallAttemptComplete.Result.Error != nil {
+			fmt.Fprintln(w, "Error:")
+			describeError(opts, opts.indent(w), kittehs.Must1(ProgramErrorFromProto(m.CallAttemptComplete.Result.Error)))
+		} else {
+			fmt.Fprintln(w, "Return Value: ")
+			describeValue(opts, opts.indent(w), kittehs.Must1(ValueFromProto(m.CallAttemptComplete.Result.Value)))
+		}
+	default:
+		fmt.Fprintln(b, "???")
```

Comment:
Please use an informative error message that users can understand

Reply:
done